ocfweb/docs/views/servers.py

```diff
@@ -56,96 +61,97 @@ def english_type(self):
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='riptide',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'anthrax',
-                    'biohazard',
-                    'death',
-                    'dementors',
-                    'democracy',
-                    'firestorm',
-                    'flood',
-                    'fraud',
-                    'gridlock',
-                    'lethe',
-                    'lightning',
-                    'maelstrom',
-                    'monsoon',
-                    'pestilence',
-                    'reaper',
-                    'segfault',
-                    'supernova',
-                    'thunder',
-                    'tsunami',
-                    'vampires',
-                    'whiteout',
-                    'zombies',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='corruption',
-            type='server',
-        ),
-
-        Host.from_ldap(
-            hostname='dev-fallingrocks',
-            type='server',
-        ),
-
-        Host('blackhole', 'network', 'Managed Arista 7050SX-64 Switch.', []),
-
-        Host('logjam', 'printer', '', []),
-        Host('pagefault', 'printer', '', []),
-        Host('papercut', 'printer', '', []),
-        Host.from_ldap('overheat', type='raspi'),
-        Host.from_ldap('tornado', type='nuc'),
-
-        *(Host.from_ldap(desktop, type='desktop')
-          for desktop in sorted(list_desktops())),
-    ]
+    servers = create_hosts(hosts_by_filter('(|(type=server)(type=desktop)(type=printer))'))
+
+    # Handle special cases
+    servers['blackhole'] = Host(
+        'blackhole', 'network',
+        'Arista 7050SX Switch.', [],
+    )
+    servers['overheat'] = servers['overheat']._replace(type='raspi')
+    servers['tornado'] = servers['tornado']._replace(type='nuc')
+
+    hypervisor_hostnames = format_query_output(query_puppet(PQL_IS_HYPERVISOR))
+    all_children = format_query_output(query_puppet(PQL_GET_CHILDREN))
+
+    # Add children to hypervisors
+    for h in list(servers.values()):
+        if h.hostname in hypervisor_hostnames:
+            # Populate a list of children
+            children = []
+            for child_hostname in all_children.get(h.hostname, []):
+                child = servers.get(child_hostname)
+                if child:
+                    del servers[child.hostname]
+                    children.append(child._replace(type='vm'))
+            # Associate host with its children and specify type
+            del servers[h.hostname]
+            servers[h.hostname] = Host(
+                hostname=h.hostname,
+                type='hypervisor',
+                description=h.description,
+                children=children,
+            )
+    return sorted(servers.values())
```

Comment:
If we don't mutate the dictionary here we would end up with two extra lists: one corresponding to the `servers` dictionary in the code snippet you posted and another list to keep track of the vms we've processed so far so we won't add them twice. After all, we don't want to just display hypervisors and their vms but also other types of machines. I agree that mutating dictionaries is ugly, but I think the alternative would require too much more code and is less efficient.

Reply:
mutating the dictionary is a classic case of premature optimization. the performance and efficiency implications of the additional list are microscopic (especially compared to the amount of CPU time it will take to query puppetdb and interpret the output), and it will probably get garbage collected anyways once the declaration goes out of scope. The code in the snippet I posted should be all that is necessary to implement this feature. Please reconsider the mutation approach.

Reply:
Mutation is also a lot harder to reason through. So +1 to abizer's suggestion.

Reply:
@abizer does your code correctly handle servers which are neither hypervisors nor VMs nor special cases?

Reply:
what servers would those be? I have not tested that snippet, obviously, but I think it should cover the cases we have.

Reply:
@abizer For now I think we have Physical Servers and Printers that are in that category.

Reply:
non-hypervisor physical machines are special cases, as are the three printers. They should be handled as such. Physical machines that aren't hypervisors (i.e. expressing the ocf_kvm class) won't even show up in the hypervisor query anyways, so it shouldn't be a problem.

Reply:
@abizer Here's a snippet implementing the special case handling you mentioned.
```python
hypervisors_hostnames = dict(format_query_output(item) for item in query_puppet(PQL_IS_HYPERVISOR))
all_children = dict(format_query_output(item) for item in query_puppet(PQL_GET_VMS))

hostnames_seen = []
servers_to_display = []
# Add children to hypervisors
for hypervisor_hostname in hypervisors_hostnames:
    children = []
    for child_hostname in all_children.get(hypervisor_hostname, []):
        child = servers.get(child_hostname)
        if child:
            children.append(child._replace(type='vm'))
            hostnames_seen.append(child.hostname)
    description = servers[hypervisor_hostname].description if hypervisor_hostname in servers else None
    servers_to_display.append(Host(
        hostname=hypervisor_hostname,
        type='hypervisor',
        description=description,
        children=children,
    ))
    hostnames_seen.append(hypervisor_hostname)

# Handle special cases
for host in servers.values():
    if (host.hostname not in hostnames_seen):
        servers_to_display.append(host)

servers_to_display.extend([
    Host(
        'blackhole', 'network',
        'Arista 7050SX Switch.', [],
    ),
    servers['overheat']._replace(type='raspi'),
    servers['tornado']._replace(type='nuc'),
])

return sorted(servers_to_display)
```